aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -29,52 +29,71 @@ Tensor _segment_reduce_cpu_kernel(
     const Tensor& lengths,
     int64_t axis,
     const c10::optional<Scalar>& initial) {
-  int64_t batch_size = lengths.numel();
-  auto output = at::empty({batch_size}, data.options());
+  int64_t segment_count = lengths.numel();
+  auto output_shape = data.sizes().vec();
+  output_shape[0] = segment_count;
+  auto output = at::empty(output_shape, data.options());
 
+  int64_t seg_element_count = output_shape.size() > 1 ? output_shape.back() : 1;
+  int64_t stride_count = data.numel() / (seg_element_count * data.size(0));
   const auto* lengths_data = lengths.data_ptr<int64_t>();
 
   AT_DISPATCH_ALL_TYPES_AND2(
       kBFloat16, kHalf, data.scalar_type(), "_segment_reduce_cpu", ([&]() {
         auto* output_data = output.data_ptr<scalar_t>();
         const auto* values_data = data.data_ptr<scalar_t>();
-        int64_t k = 0;
-        for (int64_t i = 0; i < batch_size; ++i) {
-          // ===== step1: initialize starting value
-          scalar_t initial_value;
-          if (initial.has_value()) {
-            initial_value = initial.value().to<scalar_t>();
-          } else if (reduction == SegmentReductionType::MAX) {
-            initial_value = std::numeric_limits<scalar_t>::lowest();
-          } else if (reduction == SegmentReductionType::MEAN) {
-            initial_value = 0;
-          }
+        int64_t lengths_cum_sum = 0;
+        for (int64_t i = 0; i < segment_count; ++i) {
+          for (int64_t l = 0; l < stride_count; ++l) {
```

Comment:
why are you splitting stride_count and seg_element_count into 2 loops? They are conceptually the same, the dimensions that are not being reduced.